### PR TITLE
the desktop version of the Briar secure messager

### PR DESCRIPTION
**Pull request by beroal**

### CODE REVIEW

File: apparmor.d/profiles-a-f/briar-desktop [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  owner @{HOME}/.skiko/@{hex64}/{,libskiko-*.so,skiko[0-9]*} mrw,
+
+  owner @{user_pictures_dirs}/{,**} r,
+
+  owner @{user_cache_dirs}/JNA/{,**} mrw,
+
+  owner @{tmp}/hsperfdata_@{user}/ rw,
+  owner @{tmp}/hsperfdata_@{user}/@{pid} rwk,
+  owner @{tmp}/imageio[0-9]*.tmp rw,
+  owner @{tmp}/jna[0-9]*.tmp mrw,
```
Comment (roddhjav) [inline apparmor.d/profiles-a-f/briar-desktop:59]: Use `@{int}` instead of `[0-9]*`

File: apparmor.d/profiles-a-f/briar-desktop-tor [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+  network inet stream,
+  network inet6 stream,
+  network netlink raw,
+
+  owner @{HOME}/.briar/desktop/tor/obfs4proxy Cx -> obfs4proxy,
+  owner @{HOME}/.briar/desktop/tor/snowflake Cx -> snowflake,
+
+  signal send set=term peer=briar-desktop-tor//obfs4proxy,
+  signal send set=term peer=briar-desktop-tor//snowflake,
```
Comment (roddhjav) [inline apparmor.d/profiles-a-f/briar-desktop-tor:20]: `signal` rule must come after the `network` rules

File: apparmor.d/profiles-a-f/briar-desktop [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  include <abstractions/gnome-strict>
+  include <abstractions/graphics>
+  include <abstractions/nameservice-strict>
+
+  network inet dgram,
+  network inet stream,
+  network inet6 dgram,
+  network inet6 stream,
+
+  @{lib}/jvm/java*/lib/jspawnhelper Cx -> jspawnhelper,
```
Comment (roddhjav) [inline apparmor.d/profiles-a-f/briar-desktop:24]: Move this rule with the other '@{lib}/...` rules

File: apparmor.d/profiles-a-f/briar-desktop [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  @{exec_path} mr,
+
+  @{lib}/jvm/java*/bin/java rix,
+  @{lib}/jvm/java*/lib/** rm,
+  @{sh_path} mr,
+
+  @{system_share_dirs}/icons/{,**} r,
+  @{system_share_dirs}/java/briar-desktop.jar r,
+
+  /etc/java*/{{logging,net,sound}.properties,security/java.{security,policy}} r,
```
Comment (roddhjav) [inline apparmor.d/profiles-a-f/briar-desktop:38]: I think it is safe to sum up: `/etc/java*/{,**}`

File: apparmor.d/profiles-a-f/briar-desktop [hunk 5]
```diff
@@ -0,0 +1,97 @@
+# apparmor.d - Full set of apparmor profiles
+# Copyright (C) 2024 Roman Beslik <me@beroal.in.ua>
+# SPDX-License-Identifier: GPL-2.0-only
+
+abi <abi/3.0>,
```
Comment (roddhjav) [inline apparmor.d/profiles-a-f/briar-desktop:5]: I have just updated to abi 4. Therefore, from now on:
```
abi <abi/4.0>,
```

File: apparmor.d/profiles-a-f/briar-desktop [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  ptrace read peer=briar-desktop-tor,
+  ptrace read peer=@{profile_name}//jspawnhelper,
+
+  @{exec_path} mr,
+
+  @{lib}/jvm/java*/bin/java rix,
+  @{lib}/jvm/java*/lib/** rm,
+  @{sh_path} mr,
+
+  @{system_share_dirs}/icons/{,**} r,
```
Comment (roddhjav) [inline apparmor.d/profiles-a-f/briar-desktop:35]: It  should be included in `gnome-strict`.

File: apparmor.d/profiles-a-f/briar-desktop-tor [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+abi <abi/3.0>,
+
+include <tunables/global>
+
+profile briar-desktop-tor {
+  include <abstractions/base>
+
+  network inet stream,
+  network inet6 stream,
```
Comment (roddhjav) [inline apparmor.d/profiles-a-f/briar-desktop-tor:13]: It seems unlikely that inet6 was required: tor does not work over ipv6.

File: apparmor.d/profiles-a-f/briar-desktop (hunk 1 above)
Comment (beroal) [reply apparmor.d/profiles-a-f/briar-desktop:59]: I tried. The number is too long.

Comment (roddhjav) [reply apparmor.d/profiles-a-f/briar-desktop:59]: How long? https://apparmor.pujol.io/variables/#basic-variables-of-a-given-length

Comment (beroal) [reply apparmor.d/profiles-a-f/briar-desktop:59]: 19-20 digits. I try to match the numbers with `@{int64}`, but it doesn't work for an unknown reason. The log:

```
type=AVC msg=audit(1728039823.562:27797): apparmor="DENIED" operation="unlink" class="file" profile="briar-desktop" name="/tmp/imageio18417488744924248646.tmp" pid=44544 comm="Finalizer" requested_mask="d" denied_mask="d" fsuid=1020 ouid=1020FSUID="test58" OUID="test58"
type=AVC msg=audit(1728039823.562:27798): apparmor="DENIED" operation="unlink" class="file" profile="briar-desktop" name="/tmp/imageio16063985725205008127.tmp" pid=44544 comm="Finalizer" requested_mask="d" denied_mask="d" fsuid=1020 ouid=1020FSUID="test58" OUID="test58"
type=AVC msg=audit(1728039823.562:27799): apparmor="DENIED" operation="unlink" class="file" profile="briar-desktop" name="/tmp/imageio3075731188056589370.tmp" pid=44544 comm="Finalizer" requested_mask="d" denied_mask="d" fsuid=1020 ouid=1020FSUID="test58" OUID="test58"
type=AVC msg=audit(1728039873.916:27806): apparmor="DENIED" operation="mknod" class="file" profile="briar-desktop" name="/tmp/imageio6276895853312334707.tmp" pid=44544 comm="Thread-13" requested_mask="c" denied_mask="c" fsuid=1020 ouid=1020FSUID="test58" OUID="test58"
type=AVC msg=audit(1728039880.416:27807): apparmor="DENIED" operation="mknod" class="file" profile="briar-desktop" name="/tmp/imageio15645067652751410968.tmp" pid=44544 comm="Thread-14" requested_mask="c" denied_mask="c" fsuid=1020 ouid=1020FSUID="test58" OUID="test58"
type=AVC msg=audit(1728039886.966:27808): apparmor="DENIED" operation="mknod" class="file" profile="briar-desktop" name="/tmp/imageio5235771028765679424.tmp" pid=44544 comm="Thread-15" requested_mask="c" denied_mask="c" fsuid=1020 ouid=1020FSUID="test58" OUID="test58"
type=AVC msg=audit(1728039891.983:27809): apparmor="DENIED" operation="mknod" class="file" profile="briar-desktop" name="/tmp/imageio4098760449481265574.tmp" pid=44544 comm="Thread-16" requested_mask="c" denied_mask="c" fsuid=1020 ouid=1020FSUID="test58" OUID="test58"
type=AVC msg=audit(1728039945.737:27818): apparmor="DENIED" operation="mknod" class="file" profile="briar-desktop" name="/tmp/imageio6588868399424867249.tmp" pid=44544 comm="Thread-17" requested_mask="c" denied_mask="c" fsuid=1020 ouid=1020FSUID="test58" OUID="test58"
```

File: apparmor.d/profiles-a-f/briar-desktop (hunk 3 above)
Comment (roddhjav) [reply apparmor.d/profiles-a-f/briar-desktop:24]: No, `change_profile` is a rule by itself, while `@{lib}/.. Cx` is a file rule even if it is about changing profile.

File: apparmor.d/profiles-a-f/briar-desktop (hunk 1 above)
Comment (beroal) [reply apparmor.d/profiles-a-f/briar-desktop:59]: I see that `@{int64}` means “64 decimal digits”, but I need “a 64-bit unsigned number in the decimal notation”.

File: apparmor.d/profiles-a-f/briar-desktop (hunk 6 above)
Comment (beroal) [reply apparmor.d/profiles-a-f/briar-desktop:35]: All these lines?

File: apparmor.d/profiles-a-f/briar-desktop (hunk 1 above)
Comment (roddhjav) [reply apparmor.d/profiles-a-f/briar-desktop:59]: Yes, that would be `@{u64}`, witch does not exist yet. There is `@{u8}` & `@{u16}`, I need to make the other. 
Simply put `@{u64}`. the CI will fail, but I will add it. However, in apparmor notation, that is the same than `@{int}@{int}`

File: apparmor.d/profiles-a-f/briar-desktop (hunk 6 above)
Comment (roddhjav) [reply apparmor.d/profiles-a-f/briar-desktop:35]: No only the one I selected (ie line 35, github shows too much)

File: apparmor.d/profiles-a-f/briar-desktop (hunk 1 above)
Comment (beroal) [reply apparmor.d/profiles-a-f/briar-desktop:59]: Not quite. `@{int}@{int}` matches at least two digits.

File: apparmor.d/profiles-a-f/briar-desktop (hunk 5 above)
Comment (beroal) [reply apparmor.d/profiles-a-f/briar-desktop:5]: Compiling [the most recent version of AppArmor](https://aur.archlinux.org/packages/apparmor-git) failed. I don't know where I can get the version of AppArmor supporting ABI 4.0. 😢

File: apparmor.d/profiles-a-f/briar-desktop (hunk 1 above)
Comment (roddhjav) [reply apparmor.d/profiles-a-f/briar-desktop:59]: True, but when you need more than 10 digits. It is usually not an issue (the definition of `@{u*}` are painful, and not as strict as they should be. See [18a7151](https://github.com/roddhjav/apparmor.d/commit/18a71512a9519585216f6fec11b4b7371630990f)

File: apparmor.d/profiles-a-f/briar-desktop-tor (hunk 7 above)
Comment (beroal) [reply apparmor.d/profiles-a-f/briar-desktop-tor:13]: Snowflake requires `network inet6 dgram`. I added `inet6` everywhere to make the profiles future-proof. My internet provider doesn't provide IPv6 actually.

Comment (beroal) [reply apparmor.d/profiles-a-f/briar-desktop-tor:13]: `briar-desktop` requires `network inet6 stream`.

**Comment by roddhjav** [comment]
Thanks, merged!

**Pull request merged**